src/client/pfs.go

```diff
@@ -276,15 +276,23 @@ func (c APIClient) FlushCommit(commits []*pfs.Commit, toRepos []*pfs.Repo) ([]*p
 // Blocks are content addressed and are thus identified by hashes of the content.
 // NOTE: this is lower level function that's used internally and might not be
 // useful to users.
-func (c APIClient) PutBlock(reader io.Reader) (*pfs.BlockRefs, error) {
-	putBlockClient, err := c.BlockAPIClient.PutBlock(context.Background())
+func (c APIClient) PutBlock(delimiter pfs.Delimiter, reader io.Reader) (blockRefs *pfs.BlockRefs, retErr error) {
+	writer, err := c.newPutBlockWriteCloser(delimiter)
 	if err != nil {
 		return nil, sanitizeErr(err)
 	}
-	if _, err := io.Copy(protostream.NewStreamingBytesWriter(putBlockClient), reader); err != nil {
-		return nil, err
-	}
-	return putBlockClient.CloseAndRecv()
+
+	defer func() {
+		err := writer.Close()
+		if err != nil && retErr == nil {
+			retErr = err
+		} else {
+			blockRefs = writer.blockRefs
```

Comment:
@sjezewski we talked about this while we were writing it I think. The logic is slightly weird but correct I think it's correct. There's a case where `err != nil` which implies that `writer.blockRefs == nil` but `retErr != nil` in which case we'll still run the `else` block and set `blockRefs` to `nil` which is what it already was. This has confused all 3 of us at this point though so maybe we should refactor to:

```go
defer func() {
  err := writer.Close()
  if err != nil && retErr == nil {
    retErr = err
  }
  if err != nil {
    blockRefs = writer.blockRefs
  }
}()
```